File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        Stack<TreeNode> s = new Stack<>();
        List<Integer> res = new ArrayList<>();
        TreeNode curr = root;

        while(curr != null || !s.isEmpty()){
            if(curr != null){
                s.push(curr);
                curr = curr.left;
            }else{
                TreeNode right = s.peek().right;
                if(right == null){
                    TreeNode currRoot = s.pop();
                    res.add(currRoot.val);
                    while(!s.isEmpty() && currRoot == s.peek().right){
                        currRoot = s.pop();
                        res.add(currRoot.val);
                    }  
                }else{
                    curr = right;
                }
            }
        }


        return res;

    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def rightCheck(self, curr: Optional[TreeNode], parent: Optional[TreeNode]) -> bool:
        if not curr:
            return True

        return curr.val > parent.val
    
    def leftCheck(self, curr: Optional[TreeNode], parent: Optional[TreeNode]) -> bool:
        if not curr:
            return True

        return curr.val < parent.val

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        left = self.leftCheck(root.left, root)
        right = self.rightCheck(root.right, root)

        return left and right
        
        